include/sys/spa.h

```diff
@@ -219,7 +220,7 @@ typedef struct zio_cksum_salt {
  * G		gang block indicator
  * B		byteorder (endianness)
  * D		dedup
- * X		encryption (on version 30, which is not supported)
+ * X		encryption
```

Comment:
I'll try to describe this here as best as I can and I'll add a comment describing all this as well. Hopefully this will answer some of your questions below as well.

The X bit being set indicates that this block is one of 3 types. If this is a level 0 block with an encrypted object type, the block is encrypted (see `BP_IS_ENCRYPTED()`). It will use the bp layout as described in the comment. If this is a level 0 block with an UN-encrypted object type, this block is authenticated with an HMAC (see `BP_IS_AUTHENTICATED()`) and will only use the MAC words that we have taken from the checksum. Otherwise (if level > 0), this bp will use the MAC words to store a checksum-of-MACs from the level below (see `BP_HAS_INDIRECT_MAC_CKSUM()`).

For convenience in the code, `BP_IS_PROTECTED()` refers to either encrypted or authenticated blocks and `BP_USES_CRYPT()` refers to any of these 3 kinds of blocks.